src/troute-network/troute/DataAssimilation.py

```diff
@@ -115,4 +386,438 @@ def last_obs(self):
+            nudging = streamflow_da_parameters.get('streamflow_nudging', False)
+
+        usgs_timeslices_folder = data_assimilation_parameters.get('usgs_timeslices_folder',None)
+        
+        # isolate user-input parameters for reservoir data assimilation
+        reservoir_da_parameters = data_assimilation_parameters.get('reservoir_da', None)
+
+        # check if user explictly requests USGS and/or USACE reservoir DA
+        usgs_persistence  = False
+        usace_persistence = False
+        if reservoir_da_parameters:
+            usgs_persistence  = reservoir_da_parameters.get('reservoir_persistence_usgs', False)
+            usace_persistence = reservoir_da_parameters.get('reservoir_persistence_usace', False)
+            param_file   = reservoir_da_parameters.get('gage_lakeID_crosswalk_file',None)
+        
+        # check if RFC-type reservoirs are set to true
+        rfc_params = waterbody_parameters.get('rfc')
+        if rfc_params:
+            rfc_forecast = rfc_params.get('reservoir_rfc_forecasts',False)
+            param_file = rfc_params.get('reservoir_parameter_file',None)
+        else:
+            rfc_forecast = False
+        
+        level_pool_params = waterbody_parameters.get('level_pool', defaultdict(list))
+                
+                
+        #--------------------------------------------------------------------------------
+        # Assemble USGS observation data for Streamflow DA or USGS Reservoir Persistence
+        #--------------------------------------------------------------------------------
+
+        # if user requested nudging or usgs_persistence and a specified a USGS TimeSlice directory, 
+        # then build and return USGS dataframe
+        if (nudging or usgs_persistence) and usgs_timeslices_folder:
+
+            self._usgs_df = _create_usgs_df(data_assimilation_parameters, streamflow_da_parameters, run_parameters, network, da_run)
+                
+        #--------------------------------------------------------------------------------
+        # Assemble Reservoir dataframes
+        #--------------------------------------------------------------------------------
+        
+        # if any reservoir DA is turned on, load info from reservoir parameter file:
+        break_network_at_waterbodies = waterbody_parameters.get("break_network_at_waterbodies", False)
+        if not network.wbody_conn: 
+            # Turn off any further reservoir processing if the network contains no waterbodies
+            break_network_at_waterbodies = False
+        
+        if break_network_at_waterbodies:
+            if (param_file and (usgs_persistence or usace_persistence)) or (param_file and rfc_forecast):
+                waterbody_type_specified = True
+                (
+                    waterbody_types_df, 
+                    usgs_lake_gage_crosswalk, 
+                    usace_lake_gage_crosswalk
+                ) = nhd_io.read_reservoir_parameter_file(
```

Comment:
Hmm...here is a strong connection of DA back to the nhd network, which isn't what we want with these classes.  We may need to take a close look at what the purpose of this is and where the functionality is best suited to exist.

Reply:
This function just relates gage ids to reservoir ids, so I believe this is still used for ngen, no? I don't think this has anything to do with the type of network.... But I will move this function into this module so we are not relying on nhd_io.py

Reply:
What I meant here is that the way this function works, it relies on some state/info in a reservoir param file that is specific to a particular network representation (in this case, the nhd/nwm reservoir params).  What I was suggesting is refactoring this so that DA can extract that information from an instantiated Network (whether it is an `NHDNetwork`, `HYFeaturesNetwork`, `CustomNetwork`, ect...) instead of reading hard coded params from a file that require all networks to use the same file format/layout ect...